File: java/src/main/java/com/sasteval/vuln/hard/QueryExecutor.java
package com.sasteval.vuln.hard;

import com.sasteval.util.DbUtil;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

/**
 * CWE-89: SQL Injection (Cross-File Helper - Query Executor)
 * Part of a 3-file taint chain: SqlInjectionCrossFile -> QueryBuilder -> QueryExecutor.
 */
public class QueryExecutor {

    public ResultSet executeQuery(String sql) throws SQLException {
        // VULN: Executes query string without parameterization
        Connection conn = DbUtil.getConnection();
        Statement stmt = conn.createStatement();
        return stmt.executeQuery(sql);
    }
}

File: dotnet/Vuln/Hard/QueryExecutor.cs
using System.Data.SQLite;
using SastEval.Util;

namespace SastEval.Vuln.Hard;

/// <summary>
/// Helper class for SqlInjectionCrossFile - executes raw SQL query.
/// </summary>
public static class QueryExecutor
{
    public static string Execute(string sql)
    {
        var conn = DbUtil.GetConnection();

        // VULNERABLE: Executes raw SQL string without parameterization
        using var cmd = new SQLiteCommand(sql, conn);
        using var reader = cmd.ExecuteReader();

        var result = "";
        while (reader.Read())
        {
            for (int i = 0; i < reader.FieldCount; i++)
            {
                result += $"{reader.GetName(i)}={reader[i]} ";
            }
            result += "\n";
        }

        return string.IsNullOrEmpty(result) ? "No results" : result;
    }
}
